matrix_client/crypto/olm_device.py

```diff
@@ -17,20 +17,38 @@ class OlmDevice(object):
         api (MatrixHttpApi): The api object used to make requests.
         user_id (str): Matrix user ID. Must match the one used when logging in.
         device_id (str): Must match the one used when logging in.
+        signed_otk_proportion (float): Optional. The proportion of signed one-time keys we
+            should maintain on the HS compared to unsigned keys. The maximum value of
+            ``1`` means only signed keys will be uploaded, while the minimum value of
+            ``0`` means only unsigned keys. The actual amount of keys is determined at
+            runtime from the given proportion and the maximum number of one-time keys
+            we can physically hold.
     """
 
     _olm_algorithm = 'm.olm.v1.curve25519-aes-sha2'
     _megolm_algorithm = 'm.megolm.v1.aes-sha2'
     _algorithms = [_olm_algorithm, _megolm_algorithm]
 
-    def __init__(self, api, user_id, device_id):
+    def __init__(self, api, user_id, device_id, signed_otk_proportion=1):
         self.api = api
         check_user_id(user_id)
         self.user_id = user_id
         self.device_id = device_id
         self.olm_account = olm.Account()
         logger.info('Initialised Olm Device.')
         self.identity_keys = self.olm_account.identity_keys
+        # Try to maintain half the number of one-time keys libolm can hold uploaded
+        # on the HS. This is because some keys will be claimed by peers but not
+        # used instantly, and we want them to stay in libolm, until the limit is reached
+        # and it starts discarding keys, starting by the oldest.
+        target_keys_number = self.olm_account.max_one_time_keys // 2
+        if not 0 <= signed_otk_proportion <= 1:
+            raise ValueError('signed_otk_proportion must be between 0 and 1.')
+        self.otk_target_counts = {}
+        self.otk_target_counts['signed_curve25519'] = \
```

Comment:
Can we encode the difference between signed/unsigned keys in the type system somehow instead of just having two arbitrary strings?

Reply:
Now I see that these strings are actually used in the upload request. Still feels weird to encode the difference as string keys in a dict for some reason.

Reply:
Probably the "proper" way of doing it would be to have lightweight classes for this, and move most of the logic for manipulating an individual key to the class methods.

Reply:
Yes, this is for consistency with `self.one_time_key_counts`.